### PR TITLE
Added CHOOSE operator for sets

### PR DESCRIPTION
**Pull request by mudathirmahgoub**
This PR enables THEORY_UF by default for sets and adds the operator CHOOSE for sets which returns an element from a given set. The semantics is as follows:
   * If a set A = {x}, then the term (choose A) is equivalent to the term x.
   * If the set is empty, then (choose A) is an arbitrary value.
   * If the set has cardinality > 1, then (choose A) will deterministically return an element in A.

### CODE REVIEW

File: test/regress/regress1/sets/choose3.smt2 [hunk 1]
```diff
@@ -0,0 +1,7 @@
+(set-logic ALL)
+(set-info :status unsat)
+(set-option :produce-models true)
+(declare-fun A () (Set Int))
+(assert (= (choose A) 10))
+(assert (= A (as emptyset (Set Int))))
```
Comment (mudathirmahgoub) [inline test/regress/regress1/sets/choose3.smt2:6]: Initially I thought this benchmark should answer `sat`. Then I changed my mind guessing most users would prefer this to be `unsat` and changed the semantics accordingly. Do you agree on this behavior?

File: test/regress/regress1/sets/choose4.smt2 [hunk 2]
```diff
@@ -0,0 +1,12 @@
+(set-logic ALL)
+(set-info :status sat)
+(set-option :produce-models true)
+(declare-fun A () (Set Int))
+(declare-fun a () Int)
+(assert (not (= A (as emptyset (Set Int)))))
+(assert (member 10 A))
+; this line raises an assertion error
+(assert (= a (choose A)))
```
Comment (mudathirmahgoub) [inline test/regress/regress1/sets/choose4.smt2:9]: We need to the assertion failure thrown here.

File: test/regress/regress1/sets/choose3.smt2 (hunk 1 above)
Comment (ajreynol) [reply test/regress/regress1/sets/choose3.smt2:6]: This should be sat.  choose should have partial semantics when the set is empty.  See my other comment.

File: src/api/cvc4cppkind.h [hunk 3]
```diff
@@ -1852,6 +1852,18 @@ enum CVC4_PUBLIC Kind : int32_t
    *   mkTerm(Kind kind, const std::vector<Term>& children)
    */
   COMPREHENSION,
+  /**
+   * Returns an element from a given set.
+   * If a set A = {x}, then the term (choose A) is equivalent to the term x.
+   * If the set is empty, then the atom that uses (choose A) is false.
```
Comment (ajreynol) [inline src/api/cvc4cppkind.h:1858]: If the set is empty, then (choose A) is an arbitrary element.

File: src/theory/sets/theory_sets_private.cpp [hunk 4]
```diff
@@ -1485,10 +1485,55 @@ void TheorySetsPrivate::preRegisterTerm(TNode node)
...
+Node TheorySetsPrivate::expandDefinition(LogicRequest& logicRequest, Node node)
 {
-  Debug("sets-proc") << "expandDefinition : " << n << std::endl;
-  return n;
+  Debug("sets-proc") << "expandDefinition : " << node << std::endl;
+
+  if (node.getKind() == kind::CHOOSE)
+  {
+    // (choose A) is expanded as
+    // (witness ((x elementType)) (and (member x A) (= x chooseUf(A))))
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_private.cpp:1495]: It should be (and (=> (not (= A emptyset)) (member x A)) (= x chooseUf(A)))

File: src/api/cvc4cppkind.h (hunk 3 above)
Comment (mudathirmahgoub) [reply src/api/cvc4cppkind.h]: done

File: src/theory/sets/theory_sets_private.cpp (hunk 4 above)
Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_private.cpp]: done

File: test/regress/regress1/sets/choose3.smt2 (hunk 1 above)
Comment (mudathirmahgoub) [reply test/regress/regress1/sets/choose3.smt2:6]: modified to expect `sat`

File: test/regress/regress1/sets/choose4.smt2 (hunk 2 above)
Comment (ajreynol) [reply test/regress/regress1/sets/choose4.smt2:9]: This is an assertion failure during --check-model you mean?

Comment (mudathirmahgoub) [reply test/regress/regress1/sets/choose4.smt2:9]: This is related to the issue raised in [4175](https://github.com/CVC4/CVC4/pull/4175). See the results below:
```
./cvc4 choose1.smt2 
sat
(model
(define-fun A () (Set Int) (singleton 10))
(define-fun a () Int 10)
)
```
```
./cvc4 choose1.smt2 --check-models
Warning : SmtEngine::checkModel(): cannot check simplified assertion : (= (choice ((BOUND_VARIABLE_387 Int)) (ite (= A (as emptyset (Set Int))) (= BOUND_VARIABLE_387 (chooseUf271 A)) (and (member BOUND_VARIABLE_387 A) (= BOUND_VARIABLE_387 (chooseUf271 A))))) 10)
Warning : SmtEngine::checkModel(): cannot check simplified assertion : (= (choice ((BOUND_VARIABLE_398 Int)) (ite (= A (as emptyset (Set Int))) (= BOUND_VARIABLE_398 (chooseUf271 A)) (and (member BOUND_VARIABLE_398 A) (= BOUND_VARIABLE_398 (chooseUf271 A))))) 10)
Warning : SmtEngine::checkModel(): cannot check simplified assertion : (= (choice ((BOUND_VARIABLE_413 Int)) (ite (= A (as emptyset (Set Int))) (= BOUND_VARIABLE_413 (chooseUf271 A)) (and (member BOUND_VARIABLE_413 A) (= BOUND_VARIABLE_413 (chooseUf271 A))))) 10)
sat
(model
(define-fun A () (Set Int) (singleton 10))
(define-fun a () Int 10)
)
```
```
./cvc4 choose4.smt2 
sat
(model
(define-fun A () (Set Int) (singleton 10))
Fatal failure within CVC4::Expr CVC4::SmtEngine::getValue(const CVC4::Expr&) const at /home/mudathir/Desktop/CVC4/mudathirmahgoub/src/smt/smt_engine.cpp:2986
Check failure

 m->hasApproximations() || resultNode.getKind() == kind::LAMBDA || resultNode.isConst()

Aborted (core dumped)

```

Comment (ajreynol) [reply test/regress/regress1/sets/choose4.smt2:9]: Can you add `--no-check-models` to the command line options for these regressions?  This is a common issue.

Comment (mudathirmahgoub) [reply test/regress/regress1/sets/choose4.smt2:10]: Done. The tests pass for now. But I guess a permanent solution is needed for applications like alloy.

File: src/theory/logic_info.cpp [hunk 5]
```diff
@@ -540,6 +540,12 @@ void LogicInfo::setLogicString(std::string logicString)
     IllegalArgument(logicString, err.str().c_str());
   }
 
+  if (d_theories[THEORY_SETS])
```
Comment (ajreynol) [inline src/theory/logic_info.cpp:543]: This code should not be necessary.  Can you try deleting it?

Comment (mudathirmahgoub) [reply src/theory/logic_info.cpp]: done

**Pull request merged**